### PR TITLE
#19832 : Handling exact matches for legacy IDs, which are simple numb…

### PR DESCRIPTION
**Pull request by jcastro-dotcms**
Some clients coming from old version still having legacy id, such as "1234", "1", "456" etc aka non uuid.
The fix introduce an implementation for clients that needs to support these ids and the new ones (uuid) but it is an optional implementation to avoid the decrement of the performance in clients that do not need this functionality.

Clients with legacy ids, can include this line on their dotmarketing config (or by env property)
```
dotshortyapi_use_legacy=true
```

### CODE REVIEW

**Comment by wezell** [comment]
We need to understand the performance implications of this change - every request hits these methods.

**Review by erickgonzalez — APPROVED** [review]

File: dotCMS/src/main/java/com/dotcms/uuid/shorty/ShortyIdAPIImpl.java [hunk 1]
```diff
@@ -49,35 +59,33 @@ public long getDbHits() {
...
-
-    
     try {
       if(shortStr.startsWith(TempFileAPI.TEMP_RESOURCE_PREFIX) && APILocator.getTempFileAPI().isTempResource(shortStr)) {
           return Optional.of(new ShortyId(shortStr, shortStr, ShortType.TEMP_FILE, ShortType.TEMP_FILE));
       }
-      
-      validShorty(shortStr);
-      ShortyId shortyId = null;
+      final boolean isExactMatch = NumberUtils.isParsable(shortStr) ? Boolean.TRUE : Boolean.FALSE;
```
Comment (nollymar) [inline dotCMS/src/main/java/com/dotcms/uuid/shorty/ShortyIdAPIImpl.java:66]: This logic should have a test

**Review by nollymar — CHANGES_REQUESTED** [review]

File: dotCMS/src/integration-test/java/com/dotcms/uuid/shorty/LegacyShortyIdApiTest.java [hunk 2]
```diff
@@ -0,0 +1,572 @@
...
+		builder.add(new String[] { res.get(0).get("identifier").toString(), "identifier", "template" });
+		dc.setSQL(GET_INODE, 1);
+		dc.addParam("template");
+		res = dc.loadObjectResults();
+		builder.add(new String[] { res.get(0).get("inode").toString(), "inode", "template" });
+		this.expectedIdsFromStarter = builder.build();
+	}
+
+    @Test
+    public void testShortify() {
```
Comment (dsilvam) [inline dotCMS/src/integration-test/java/com/dotcms/uuid/shorty/LegacyShortyIdApiTest.java:261]: javadoc

File: dotCMS/src/integration-test/java/com/dotcms/uuid/shorty/LegacyShortyIdApiTest.java [hunk 3]
```diff
@@ -0,0 +1,572 @@
...
+
+        final String shortId12 = "12345-6789-012";
+        Assert.assertEquals(shortyIdAPI.shortify(shortId12), "1234567890");
+
+        final String shortId14 = "12345-6789-012-11";
+        Assert.assertEquals(shortyIdAPI.shortify(shortId14), "1234567890");
+    }
+
+	@Test
+    public void testShortyLookup() {
```
Comment (dsilvam) [inline dotCMS/src/integration-test/java/com/dotcms/uuid/shorty/LegacyShortyIdApiTest.java:296]: javadoc

File: dotCMS/src/integration-test/java/com/dotcms/uuid/shorty/LegacyShortyIdApiTest.java [hunk 4]
```diff
@@ -0,0 +1,572 @@
...
+                System.out.println("bad shorty:" + key);
+                System.out.println("looking shortType:" + shortType);
+                System.out.println("looking shortSubType:" + shortSubType);
+                throw t;
+            }
+        }
+    }
+
+    @Test
+    public void testShortyCache() {
```
Comment (dsilvam) [inline dotCMS/src/integration-test/java/com/dotcms/uuid/shorty/LegacyShortyIdApiTest.java:338]: javadoc

File: dotCMS/src/integration-test/java/com/dotcms/uuid/shorty/LegacyShortyIdApiTest.java [hunk 5]
```diff
@@ -0,0 +1,572 @@
...
+            String key = api.shortify(val[0]);
+            Optional<ShortyId> opt = api.getShorty(key);
+        }
+
+        long dbQueries3 = api.getDbHits();
+        assert (dbQueries3 == dbQueries2 + expectedIds.size());
+    }
+
+    @Test
+    public void testValidShorty() {
```
Comment (dsilvam) [inline dotCMS/src/integration-test/java/com/dotcms/uuid/shorty/LegacyShortyIdApiTest.java:379]: javadoc

File: dotCMS/src/integration-test/java/com/dotcms/uuid/shorty/LegacyShortyIdApiTest.java [hunk 6]
```diff
@@ -0,0 +1,572 @@
...
+                api.validShorty(x);
+                assert (true);
+            } catch (Exception e) {
+                assert (false);
+            }
+        }
+    }
+
+    @Test
+    public void test404Cache() {
```
Comment (dsilvam) [inline dotCMS/src/integration-test/java/com/dotcms/uuid/shorty/LegacyShortyIdApiTest.java:411]: javadoc

File: dotCMS/src/integration-test/java/com/dotcms/uuid/shorty/LegacyShortyIdApiTest.java [hunk 7]
```diff
@@ -0,0 +1,572 @@
...
+        for (String key : fourOhFours) {
+            Optional<ShortyId> opt = api.getShorty(key);
+        }
+
+        long dbQueries3 = api.getDbHits();
+        assert (dbQueries3 == dbQueries2 + fourOhFours.size());
+    }
+
+    @Test
+    public void testIdentifier404CacheInvalidation() throws Exception{
```
Comment (dsilvam) [inline dotCMS/src/integration-test/java/com/dotcms/uuid/shorty/LegacyShortyIdApiTest.java:443]: javadoc

File: dotCMS/src/integration-test/java/com/dotcms/uuid/shorty/LegacyShortyIdApiTest.java [hunk 8]
```diff
@@ -0,0 +1,572 @@
...
+        
+        for(int i=testUuid.length();i>ShortyIdAPIImpl.MINIMUM_SHORTY_ID_LENGTH;i--) {
+            String test=testUuid.substring(0, i);
+            assert(api.getShorty(test).isPresent());
+        }
+        
+    }
+    
+    @Test
+    public void testContentlet404CacheInvalidation() throws Exception{
```
Comment (dsilvam) [inline dotCMS/src/integration-test/java/com/dotcms/uuid/shorty/LegacyShortyIdApiTest.java:478]: javadoc

File: dotCMS/src/integration-test/java/com/dotcms/uuid/shorty/LegacyShortyIdApiTest.java [hunk 9]
```diff
@@ -0,0 +1,572 @@
...
+            String test=testUuid.substring(0, i);
+            assert(api.getShorty(test).isPresent());
+        }
+        
+    }
+    
+    
+    
+    @Test
+    public void testUuidIfy() {
```
Comment (dsilvam) [inline dotCMS/src/integration-test/java/com/dotcms/uuid/shorty/LegacyShortyIdApiTest.java:515]: javadoc

File: dotCMS/src/integration-test/java/com/dotcms/uuid/shorty/LegacyShortyIdApiTest.java [hunk 10]
```diff
@@ -0,0 +1,572 @@
...
+
+        for (String[] x : expectedIds) {
+            String noDashes = x[0].replaceAll("-", "");
+            String y = api.uuidIfy(noDashes);
+            assert (x[0].equals(y));
+        }
+    }
+
+    @Test
+    public void testLongerShorties() {
```
Comment (dsilvam) [inline dotCMS/src/integration-test/java/com/dotcms/uuid/shorty/LegacyShortyIdApiTest.java:531]: javadoc

File: dotCMS/src/integration-test/java/com/dotcms/uuid/shorty/LegacyShortyIdApiTest.java [hunk 11]
```diff
@@ -0,0 +1,572 @@
...
+                } catch (Throwable t) {
+                    System.out.println("key is empty:" + key);
+                    throw t;
+                }
+            }
+        }
+    }
+    
+    @Test
+    public void testTempShorties() throws DotSecurityException, IOException {
```
Comment (dsilvam) [inline dotCMS/src/integration-test/java/com/dotcms/uuid/shorty/LegacyShortyIdApiTest.java:550]: javadoc

File: dotCMS/src/integration-test/java/com/dotcms/uuid/shorty/LegacyShortyIdApiTest.java [hunk 12]
```diff
@@ -0,0 +1,377 @@
...
+
+    /**
+     * Method to test: {@link ShortyIdAPI#shortify(String)}
+     * Given Scenario: Shortify the String (will trim the - and leave it at 10 char)
+     * ExpectedResult: success all shortify
+     */
+    @Test
+    @UseDataProvider("dataProviderStringToShortify")
+    public void test_shortify(final Tuple2<String,String> testCase) {
+        Assert.assertEquals(shortyIdAPI.shortify(testCase._1), testCase._2);
```
Comment (nollymar) [inline dotCMS/src/integration-test/java/com/dotcms/uuid/shorty/LegacyShortyIdApiTest.java:216]: the expected result should be the first parameter of the assert

**Review by nollymar — APPROVED** [review]

**Pull request merged**